python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,215 @@
+        "default-sort-order-id": {"type": "integer"},
+    },
+    "required": [
+        "format-version",
+        "table-uuid",
+        "location",
+        "last-sequence-number",
+        "last-updated-ms",
+        "last-column-id",
+        "schemas",
+        "current-schema-id",
+        "partition-spec",
+        "partition-specs",
+        "default-spec-id",
+        "last-partition-id",
+        "properties",
+        "current-snapshot-id",
+        "snapshots",
+        "snapshot-log",
+        "metadata-log",
+        "sort-orders",
+        "default-sort-order-id",
+    ],
+}
+TABLE_METADATA_V2_SCHEMA = {
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "type": "object",
+    "properties": {
+        "format-version": {"type": "string"},
+        "table-uuid": {"type": "string"},
+        "location": {"type": "string"},
+        "last-sequence-number": {"type": "integer"},
+        "last-updated-ms": {"type": "integer"},
+        "last-column-id": {"type": "integer"},
+        "schemas": {"type": "array", "items": {}},
+        "current-schema-id": {"type": "integer"},
+        "partition-spec": {"type": "array", "items": {}},
+        "partition-specs": {"type": "array", "items": {}},
+        "default-spec-id": {"type": "integer"},
+        "last-partition-id": {"type": "integer"},
+        "properties": {"type": "object"},
+        "current-snapshot-id": {"type": "number"},
+        "snapshots": {"type": "array", "items": {}},
+        "snapshot-log": {"type": "array", "items": {}},
+        "metadata-log": {"type": "array", "items": {}},
+        "sort-orders": {"type": "array", "items": {}},
+        "default-sort-order-id": {"type": "integer"},
+    },
+    "required": [
+        "format-version",
+        "table-uuid",
+        "location",
+        "last-sequence-number",
+        "last-updated-ms",
+        "last-column-id",
+        "schemas",
+        "current-schema-id",
+        "partition-spec",
+        "partition-specs",
+        "default-spec-id",
+        "last-partition-id",
+        "properties",
+        "current-snapshot-id",
+        "snapshots",
+        "snapshot-log",
+        "metadata-log",
+        "sort-orders",
+        "default-sort-order-id",
+    ],
+}
+
+
+class TableMetadata:
+    """Metadata for an Iceberg table as specified in the Apache Iceberg
+    spec (https://iceberg.apache.org/spec/#iceberg-table-spec)
+
+    Args:
+      metadata (dict): Table metadata dictionary representation
+      version (str|int): The metadata spec version (1 or 2)
+    """
+
+    def __init__(self, metadata: dict, version: Union[str, int]):
+        self._version = version
+        self._metadata = metadata
+        for name, value in metadata.items():
+            setattr(self, self._clean_attribute_name(name), self._wrap(value))
+
+    def _wrap(self, value: Any):
+        """A recursive function that drills into iterable values and returns
+        nested TableMetadata instances
```

Comment:
Why would there be nested table metadata instances?

Reply:
The class name does make this sound odd but the logic is essentially a DFS through the metadata to assign everything as class attributes. If a value in the json is an object (like properties), it instantiates that as a sort of "partial" `TableMetadata` instance and then starts a new traversal through the contents of that object. In other words 
```py
table_metadata = TableMetadata(...)
isinstance(table_metadata, TableMetadata)  # True
isinstance(table_metadata.properties, TableMetadata)  # True

isinstance(table_metadata.snapshot_log, list)  # True
isinstance(table_metadata.snapshot_log[0], TableMetadata)  # True
```
If the idea of this partial metadata existing as a `TableMetadata` instance doesn't sit well, I could instead have a generic `Metadata` or `Config` class that actually traverses the json object to create a class with all of the class attributes, and have that as an argument to a `TableMetadata` class that handles validation and other table metadata related things.

Something like:
```py
config = Config({"table-uuid": "foo", ...})
table_metadata = TableMetadata(config, version="2")
```

The `_wrap()` method would live in the `Config` class and the docstring would read as:

> A recursive function that drills into iterable values and returns
        nested Config instances

Reply:
Okay, I think I get it.

I like being able to use `m.snapshot_log[0]` and similar ways to access the metadata. But, I'm not sure that this is changing enough that it's worth the wrapper approach, instead of just converting to a `TableMetadata` class that pulls out and stores `self.current_schema_id` (for example). Table metadata shouldn't be _that_ complicated since it's mostly a few lists of objects at the most nested level (like metadata > snapshots > snapshot > properties > key/value).

While this makes it easy to get started, it would be awkward to make updates to the metadata as JSON because you'd need to produce a new JSON tree and then wrap with this class. We may also want classes for things like Snapshot, which can embed some operations like reading metadata.

Reply:
I've updated the PR to remove the wrapper approach and set the top level as class attributes. I've also added a `to_dict()` method that's used by `validate()` where the `TableMetadata` instance is serialized into a python dictionary and then passed into the jsonschema validate function. I've also added freezing of instances and this punts the question of how we perform table metadata updates.

I can follow-up with a PR that uses the builder pattern where a python `dict` is the mutable form. So building from an existing table metadata instance would start with calling `to_dict()`.